core/utils/container_utils.hpp

```diff
@@ -27,339 +27,146 @@
-      : buffer_t{std::move(other)}, ptr{std::move(other.ptr)} {}
-
-    typename Allocator::value_type ptr;
-  };
-
-  static_assert(std::is_nothrow_move_constructible_v<buffer_entry_t>,
-                "default move constructor expected");
-
-  buffer_t& push_buffer(size_t offset, const bucket_size_t& bucket) {
-    buffers_.emplace_back(offset, bucket.size, alloc_.get().allocate(bucket));
-    return buffers_.back();
-  }
-
-  IRS_NO_UNIQUE_ADDRESS EboRef<allocator_type> alloc_;
-  std::vector<buffer_entry_t> buffers_;
+  IRS_NO_UNIQUE_ADDRESS std::allocator<byte_type> alloc_;
+  std::vector<buffer_t> buffers_;
 };
 
 //////////////////////////////////////////////////////////////////////////////
 /// @brief a container allowing raw access to internal storage, and
 ///        using an allocation strategy similar to an std::deque
 //////////////////////////////////////////////////////////////////////////////
-template<size_t NumBuckets, size_t SkipBits,
-         typename Allocator = memory::default_allocator>
-class raw_block_vector : public raw_block_vector_base<Allocator> {
+template<size_t NumBuckets, size_t SkipBits>
+class raw_block_vector : public raw_block_vector_base {
  public:
-  static const size_t NUM_BUCKETS = NumBuckets;  // total number of buckets
-  static const size_t FIRST_BUCKET_SIZE = 1 << SkipBits;
-
-  typedef raw_block_vector_base<Allocator> base_t;
-  typedef typename base_t::allocator_type allocator_type;
-
-  explicit raw_block_vector(const Allocator& alloc /*= Allocator()*/) noexcept
-    : base_t{alloc} {}
-
-  raw_block_vector(raw_block_vector&& other) noexcept
-    : base_t{std::move(other)} {}
+  static constexpr BucketMeta<NumBuckets, SkipBits> kMeta{};
+  static constexpr BucketInfo kLast = kMeta[NumBuckets - 1];
 
   IRS_FORCE_INLINE size_t buffer_offset(size_t position) const noexcept {
-    // non-precomputed bucket size is the same as the last precomputed bucket
-    // size
-    return position < LAST_BUFFER.offset
-             ? compute_bucket_offset<SkipBits>(position)
-             : (LAST_BUFFER_ID +
-                (position - LAST_BUFFER.offset) / LAST_BUFFER.size);
+    return position < kLast.offset
+             ? ComputeBucketOffset<SkipBits>(position)
+             : (NumBuckets - 1 + (position - kLast.offset) / kLast.size);
   }
 
-  typename base_t::buffer_t& push_buffer() {
-    if (base_t::buffers_.size() <
-        META.size()) {  // one of the precomputed buckets
-      const auto& bucket = META[base_t::buffers_.size()];
-      return base_t::push_buffer(bucket.offset, bucket);
-    }
-
-    // non-precomputed buckets, offset is the sum of previous buckets
-    IRS_ASSERT(
-      !base_t::buffers_
-         .empty());  // otherwise do not know what size buckets to create
-    const auto& bucket =
-      base_t::buffers_.back();  // most of the meta from last computed bucket
-    return base_t::push_buffer(bucket.offset + bucket.size, META.back());
+  const buffer_t& push_buffer() {
```

Comment:
why it is not in base class? Looks a bit strange that base has "clear" but not "push"

Reply:
Because it's need META

Reply:
But then the question - why we need to keep `raw_block_vector_base` ? It is not used anywhere else. Could we just get rid of it and put everything in `raw_block_vector`?

Reply:
Same as old reason: to instantiate over methods single time

Reply:
then why not with "push_buffer" part that does not depends on META ?